cpp/src/request_mem.cpp

```diff
@@ -0,0 +1,196 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include <cstdio>
+#include <memory>
+#include <string>
+
+#include <ucp/api/ucp.h>
+
+#include <ucxx/buffer.h>
+#include <ucxx/delayed_submission.h>
+#include <ucxx/request_mem.h>
+
+namespace ucxx {
+
+std::shared_ptr<RequestMem> createRequestMem(
+  std::shared_ptr<Endpoint> endpoint,
+  const std::variant<data::MemPut, data::MemGet> requestData,
+  const bool enablePythonFuture                = false,
+  RequestCallbackUserFunction callbackFunction = nullptr,
+  RequestCallbackUserData callbackData         = nullptr)
+{
+  std::shared_ptr<RequestMem> req = std::visit(
+    data::dispatch{
+      [&endpoint, &enablePythonFuture, &callbackFunction, &callbackData](data::MemPut memPut) {
+        return std::shared_ptr<RequestMem>(new RequestMem(
+          endpoint, memPut, "memPut", enablePythonFuture, callbackFunction, callbackData));
+      },
+      [&endpoint, &enablePythonFuture, &callbackFunction, &callbackData](data::MemGet memGet) {
+        return std::shared_ptr<RequestMem>(new RequestMem(
+          endpoint, memGet, "memGet", enablePythonFuture, callbackFunction, callbackData));
+      },
+    },
+    requestData);
+
+  // A delayed notification request is not populated immediately, instead it is
+  // delayed to allow the worker progress thread to set its status, and more
+  // importantly the Python future later on, so that we don't need the GIL here.
+  req->_worker->registerDelayedSubmission(
+    req, std::bind(std::mem_fn(&Request::populateDelayedSubmission), req.get()));
+
+  return req;
+}
+
+RequestMem::RequestMem(std::shared_ptr<Endpoint> endpoint,
+                       const std::variant<data::MemPut, data::MemGet> requestData,
+                       const std::string operationName,
+                       const bool enablePythonFuture,
+                       RequestCallbackUserFunction callbackFunction,
+                       RequestCallbackUserData callbackData)
+  : Request(endpoint, data::getRequestData(requestData), operationName, enablePythonFuture)
+{
+  std::visit(data::dispatch{
+               [this](data::MemPut memPut) {
+                 if (_endpoint == nullptr)
+                   throw ucxx::Error("A valid endpoint is required to send memory messages.");
+               },
+               [this](data::MemGet memGet) {
+                 if (_endpoint == nullptr)
+                   throw ucxx::Error("A valid endpoint is required to receive memory messages.");
+               },
+               [](auto) { throw std::runtime_error("Unreachable"); },
+             },
+             requestData);
+
+  _callback     = callbackFunction;
+  _callbackData = callbackData;
```

Comment:
Why defer setting these? Couldnt you just add `_callback(std::move(callbackFunction))` after the base constructor?

Reply:
The reason is this is a member of the base class, which can't be part of the constructor initializer list and thus we must add those to the base class constructor. This is the same in other `Request*` types which are on my plans to be done, but I haven't got there yet.

Reply:
I started https://github.com/rapidsai/ucxx/pull/194 to do that, leaving as draft until this PR and https://github.com/rapidsai/ucxx/pull/192 are merged because both use this pattern.